jetty-server/src/main/java/org/eclipse/jetty/server/session/NullSessionCache.java

```diff
@@ -30,6 +36,155 @@
+        }
+
+        @Override
+        public void attributeRemoved(HttpSessionBindingEvent event)
+        {
+            doAttributeChanged(event);
+        }
+
+        @Override
+        public void attributeReplaced(HttpSessionBindingEvent event)
+        {
+            doAttributeChanged(event);
+        }
+
+        private void doAttributeChanged(HttpSessionBindingEvent event)
+        {
+            if (_writeThroughMode == WriteThroughMode.ON_EXIT)
+                return;
+
+            Session session = (Session)event.getSession();
+
+            if (_writeThroughMode == WriteThroughMode.ALWAYS
+                || (_writeThroughMode == WriteThroughMode.NEW && session.isNew()))
+            {
+                //ensure that a call to willPassivate doesn't result in a passivation
+                //listener removing an attribute, which would cause this listener to
+                //be called again
+                if (_sessionsBeingWritten.add(session))
+                {
+                    SessionDataStore store = getSessionDataStore();
+                    try
+                    {
+                        if (store != null)
+                        {
+                            //should hold the lock on the session, but as sessions are never shared
+                            //with the NullSessionCache, there can be no other thread modifying the 
+                            //same session at the same time (although of course there can be another
+                            //request modifying its copy of the session data, so it is impossible
+                            //to guarantee the order of writes).
+                            if (store.isPassivating())
+                                session.willPassivate();
+                            store.store(session.getId(), session.getSessionData());
+                            if (store.isPassivating())
+                                session.didActivate();
+                        }
+                    }
+                    catch (Exception e)
+                    {
+                        LOG.warn("Write through of {} failed", e);
+                    }
+                    finally
+                    {
+                        _sessionsBeingWritten.remove(session);
+                    }
+                }
+            }
+        }
+    }
+
+
+    
+    /**
+     * ALWAYS means write through every attribute change.
+     * NEW means to write through every attribute change only 
+     * while the session is freshly created, ie its id has not yet been returned to the client.
+     * ON_EXIT means write the session only when the request exits (which is the behaviour of AbstractSessionCache).
+     *
+     */
+    public enum WriteThroughMode
+    {
+        ALWAYS,
+        NEW,
+        ON_EXIT
+    };
+    
+    
+    /**
+     * Default is to only write out the session if it is new or changed
+     * when the request exits.
+     */
```

Comment:
This javadoc should be in the `WriteThroughMode` class, or the `NullSessionCache` class.